File: atcoder-problems-backend/src/main/scala/com/kenkoooo/atcoder/model/MergedProblem.scala
package com.kenkoooo.atcoder.model

import com.kenkoooo.atcoder.common.TypeAnnotations.{ContestId, ProblemId, SubmissionId, UserId}
import scalikejdbc.{WrappedResultSet, _}

/**
  * [[Problem]] and other information
  *
  * @param id                   Problem ID
  * @param contestId            Contest ID
  * @param title                Problem title
  * @param fastestSubmissionId  id of the fastest submission of this problem
  * @param fastestUserId        user id of the fastest submission of this problem
  * @param fastestContestId     contest id of the fastest submission of this problem
  * @param executionTime        execution time of the fastest submission of this problem
  * @param firstSubmissionId    id of the first submission of this problem
  * @param firstUserId          user id of the first submission of this problem
  * @param firstContestId       contest id of the first submission of this problem
  * @param shortestSubmissionId id of the shortest submission of this problem
  * @param shortestUserId       user id of the shortest submission of this problem
  * @param shortestContestId    contest id of the shortest submission of this problem
  * @param sourceCodeLength     source code length of the shortest submission of this problem
  * @param solverCount          the number of solvers of this problem
  */
case class MergedProblem(id: ProblemId,
                         contestId: ContestId,
                         title: String,
                         fastestSubmissionId: Option[SubmissionId],
                         fastestUserId: Option[UserId],
                         fastestContestId: Option[ContestId],
                         executionTime: Option[Int],
                         firstSubmissionId: Option[SubmissionId],
                         firstUserId: Option[UserId],
                         firstContestId: Option[ContestId],
                         shortestSubmissionId: Option[SubmissionId],
                         shortestUserId: Option[UserId],
                         shortestContestId: Option[ContestId],
                         sourceCodeLength: Option[Int],
                         solverCount: Int)

object MergedProblem extends SQLSyntaxSupport[MergedProblem] {
  def apply(problem: SyntaxProvider[Problem],
            fastestSubmission: SyntaxProvider[Submission],
            firstSubmission: SyntaxProvider[Submission],
            shortestSubmission: SyntaxProvider[Submission],
            solver: SyntaxProvider[ProblemSolver])(rs: WrappedResultSet): MergedProblem =
    MergedProblem(
      id = rs.string(problem.resultName.id),
      contestId = rs.string(problem.resultName.contestId),
      title = rs.string(problem.resultName.title),
      fastestSubmissionId = rs.longOpt(fastestSubmission.resultName.id),
      fastestUserId = rs.stringOpt(fastestSubmission.resultName.userId),
      fastestContestId = rs.stringOpt(fastestSubmission.resultName.contestId),
      executionTime = rs.intOpt(fastestSubmission.resultName.executionTime),
      firstSubmissionId = rs.longOpt(firstSubmission.resultName.id),
      firstUserId = rs.stringOpt(firstSubmission.resultName.userId),
      firstContestId = rs.stringOpt(firstSubmission.resultName.contestId),
      shortestSubmissionId = rs.longOpt(shortestSubmission.resultName.id),
      shortestUserId = rs.stringOpt(shortestSubmission.resultName.userId),
      shortestContestId = rs.stringOpt(shortestSubmission.resultName.contestId),
      sourceCodeLength = rs.intOpt(shortestSubmission.resultName.length),
      solverCount = rs.intOpt(solver.resultName.userCount).getOrElse(0)
    )
}

File: atcoder-problems-frontend/src/interfaces/MergedProblem.ts
import {
  hasPropertyAsType,
  hasPropertyAsTypeOrNull,
  hasPropertyAsTypeOrUndefined,
  isNumber,
  isString,
} from "../utils/TypeUtils";

// eslint-disable-next-line import/no-default-export
export default interface MergedProblem {
  // Basic information
  readonly id: string;
  readonly contest_id: string;
  readonly title: string;

  // Information for first AC
  readonly first_user_id: string | null;
  readonly first_contest_id: string | null;
  readonly first_submission_id: number | null;

  // Information for fastest code
  readonly fastest_user_id: string | null;
  readonly fastest_contest_id: string | null;
  readonly fastest_submission_id: number | null;
  readonly execution_time: number | null;

  // Information for shortest code
  readonly shortest_user_id: string | null;
  readonly shortest_contest_id: string | null;
  readonly shortest_submission_id: number | null;
  readonly source_code_length: number | null;

  readonly solver_count: number | null;
  readonly point?: number | null;
}

export const isMergedProblem = (obj: unknown): obj is MergedProblem =>
  typeof obj === "object" &&
  obj !== null &&
  hasPropertyAsType(obj, "id", isString) &&
  hasPropertyAsType(obj, "contest_id", isString) &&
  hasPropertyAsType(obj, "title", isString) &&
  hasPropertyAsTypeOrNull(obj, "first_user_id", isString) &&
  hasPropertyAsTypeOrNull(obj, "first_contest_id", isString) &&
  hasPropertyAsTypeOrNull(obj, "first_submission_id", isNumber) &&
  hasPropertyAsTypeOrNull(obj, "fastest_user_id", isString) &&
  hasPropertyAsTypeOrNull(obj, "fastest_contest_id", isString) &&
  hasPropertyAsTypeOrNull(obj, "fastest_submission_id", isNumber) &&
  hasPropertyAsTypeOrNull(obj, "execution_time", isNumber) &&
  hasPropertyAsTypeOrNull(obj, "shortest_user_id", isString) &&
  hasPropertyAsTypeOrNull(obj, "shortest_contest_id", isString) &&
  hasPropertyAsTypeOrNull(obj, "shortest_submission_id", isNumber) &&
  hasPropertyAsTypeOrNull(obj, "source_code_length", isNumber) &&
  hasPropertyAsTypeOrNull(obj, "solver_count", isNumber) &&
  (hasPropertyAsTypeOrUndefined(obj, "point", isNumber) ||
    hasPropertyAsTypeOrNull(obj, "solver_count", isNumber));
